File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        if p.val > q.val:
            temp = q
            q = p
            p = temp
        if p.val <= root.val <= q.val:
            return root
        
        if root.val < p.val and root.val < q.val and root.right is not None:
            return self.lowestCommonAncestor(root.right, p, q)
        else:
            return self.lowestCommonAncestor(root.left, p, q)

File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
    unordered_map<int, int> inorderIdx{};
public:
    TreeNode* build(const vector<int>& preorder, int preL, int preR,
    const vector<int>& inorder, int inL, int inR) {
        if (preL == preR) return nullptr;
        auto rootVal{preorder[preL]};
        TreeNode* root{new TreeNode(rootVal)};
        auto rootValInorderIdx{inorderIdx[rootVal]};
        auto leftSubtreeSize{rootValInorderIdx - inL};
        auto rightSubtreeSize{inR - rootValInorderIdx + 1};
        root->left = build(preorder, preL + 1, preL + 1 + leftSubtreeSize, inorder, inL, rootValInorderIdx);
        root->right = build(preorder, preL + 1 + leftSubtreeSize, preR, inorder, rootValInorderIdx + 1, inR);
        return root;
    }

    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        for (auto i{0}; i < preorder.size(); ++i) {
            inorderIdx[inorder[i]] = i;
        }
        return build(preorder, 0, preorder.size(), inorder, 0, inorder.size());
    }
};
